modules/apps/frontend-js/frontend-js-react-web/src/main/resources/META-INF/resources/js/components/A11y/A11yChecker.ts

```diff
@@ -0,0 +1,398 @@
+	}
+}
+
+/**
+ * Scheduler deals with scheduling a callback to A11y that runs axe-core at the
+ * end, many of the logics and algorithms in the scheduler are restricted to
+ * A11y but partially generic.
+ *
+ * The implementation is based on Queue. Each new scheduled callback is added
+ * end of the queue, the scheduler works with the main callback which is
+ * responsible for emptying the queue and executing the queue callbacks in the
+ * work loop respecting the frame deadline.
+ *
+ * Main thread
+ * ------------------------------------------------------------------------->
+ *
+ * +------------+ ++ ++ ++ +--------+            +--------+    +------+
+ * |    Busy    | || || || |  Idle  |            |  Busy  |    | Idle |
+ * +------------+ ++ ++ ++ +--------+            +--------+    +------+
+ *
+ *                         +----------+ +---+                  +-----------+
+ *                         | Job Exec.| |   |   Yield to host  | Job Exec. |
+ *                         +----------+ +---+ <--------------> +-----------+
+ *                        |                  |
+ *                        +------------------+
+ *                              Deadline
+ *
+ * When the main thread is idle the main callback is called to empty the queue
+ * then a work loop is executed to ensure that we can execute more than one
+ * callback until the defined deadline but it is interrupted to avoid loss
+ * of frame and at the end of each execution it defines the next frame phase
+ * that avoids abusing cycles and yields up control of the main thread to
+ * the host. The main callback will be rescheduled again until the queue is
+ * empty.
+ */
+class Scheduler<T> {
+	private taskIdCounter: number = 0;
+	private readonly queue: Queue<Task<T>>;
+	private currentTask: Task<T> | undefined = undefined;
+	private scheduledHostHandle: number = 0;
+	private isCallbackScheduled: boolean = false;
+	private isPerformingWork: boolean = false;
+
+	/**
+	 * The frame phases can be recognized as lanes which are a space of time
+	 * where the scheduler yields up the main thread to the host in that time
+	 * interval before taking up again.
+	 */
+	private nextFramePhase: number = 0;
+
+	/**
+	 * The yield interval is defined based on the RAIL model, this value is
+	 * static and means that we yield 100ms to the host and define the deadline
+	 * to maximize the chances of hitting 60 fps.
+	 */
+	private readonly yieldInterval: number = 100;
+	private deadline: number = 0;
+
+	constructor() {
+		this.queue = new Queue();
+	}
+
+	cancelHostCallback() {
+		window.cancelIdleCallback(this.scheduledHostHandle);
+	}
+
+	hasCallback(selector: Selector<Task<T>>) {
+		return this.queue.has(selector);
+	}
+
+	private async flushWork() {
+
+		// We'll need a host callback the next time work is scheduled.
+
+		this.isCallbackScheduled = false;
+
+		this.isPerformingWork = true;
+		try {
+			return await this.workLoop();
+		}
+		finally {
+			this.currentTask = undefined;
```

Comment:
nit: more of a my own personal opinion, but I think we should avoid setting undefined as a value and use `null` instead. The way I read this is "We are defining the value of `this.currentTask` as undefined", so it's always made more sense to me that null is used when we want to define an "empty" value.

Reply:
Yeah I also have a preference for `null`, actually I went with this first but I started having problems with type, I don’t remember now but it was some conflict with the return of a native function that was undefined. I probably don't need to use `undefined` and I can use `null` now.